Mathlib/Combinatorics/Additive/Corner/Roth.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import Mathlib.Combinatorics.Additive.AP.Three.Defs
+import Mathlib.Combinatorics.Additive.Corner.Defs
+import Mathlib.Combinatorics.SimpleGraph.Triangle.Removal
+import Mathlib.Combinatorics.SimpleGraph.Triangle.Tripartite
+
+/-!
+# The corners theorem and Roth's theorem
+
+This file proves the corners theorem and Roth's theorem on arithmetic progressions of length three.
+
+## References
+
+* [Yaël Dillies, Bhavik Mehta, *Formalising Szemerédi’s Regularity Lemma in Lean*][srl_itp]
+* [Wikipedia, *Corners theorem*](https://en.wikipedia.org/wiki/Corners_theorem)
+-/
+
+open Finset SimpleGraph TripartiteFromTriangles
+open Function hiding graph
+open Fintype (card)
+
+variable {G : Type*} [AddCommGroup G] [Fintype G] [DecidableEq G] {A B : Finset (G × G)}
+  {a b c d x y : G} {n : ℕ} {ε : ℝ}
+
+namespace Corners
+
+/-- The triangle indices for the proof of the corners theorem construction. -/
+private def triangleIndices (A : Finset (G × G)) : Finset (G × G × G) :=
+  A.map ⟨fun (a, b) ↦ (a, b, a + b), by rintro ⟨x₁, x₂⟩ ⟨y₁, y₂⟩ ⟨⟩; rfl⟩
+
+@[simp]
+private lemma mk_mem_triangleIndices : (a, b, c) ∈ triangleIndices A ↔ (a, b) ∈ A ∧ c = a + b := by
+  simp only [triangleIndices, Prod.ext_iff, mem_map, Embedding.coeFn_mk, exists_prop, Prod.exists,
+    eq_comm]
+  refine ⟨?_, fun h ↦ ⟨_, _, h.1, rfl, rfl, h.2⟩⟩
+  rintro ⟨_, _, h₁, rfl, rfl, h₂⟩
+  exact ⟨h₁, h₂⟩
+
+@[simp] private lemma card_triangleIndices : (triangleIndices A).card = A.card := card_map _
+
+private instance triangleIndices.instExplicitDisjoint : ExplicitDisjoint (triangleIndices A) := by
+  constructor
+  all_goals
+    simp only [mk_mem_triangleIndices, Prod.mk.inj_iff, exists_prop, forall_exists_index, and_imp]
+    rintro a b _ a' - rfl - h'
+    simp [Fin.val_eq_val, *] at * <;> assumption
+
+private lemma noAccidental (hs : IsCornerFree (A : Set (G × G))) :
+    NoAccidental (triangleIndices A) where
+  eq_or_eq_or_eq a a' b b' c c' ha hb hc := by
+    simp only [mk_mem_triangleIndices] at ha hb hc
+    exact .inl $ hs ⟨hc.1, hb.1, ha.1, hb.2.symm.trans ha.2⟩
+
+private lemma farFromTriangleFree_graph (hε : ε * card G ^ 2 ≤ A.card) :
+    (graph $ triangleIndices A).FarFromTriangleFree (ε / 9) := by
+  refine farFromTriangleFree _ ?_
+  simp_rw [card_triangleIndices, mul_comm_div, Nat.cast_pow, Nat.cast_add]
+  ring_nf
+  simpa only [mul_comm] using hε
+
+end Corners
+
+open Corners
+
+/-- The **corners theorem** for finite abelian groups.
+
+The maximum density of a corner-free set in `G × G` goes to zero as `|G|` tends to infinity. -/
+theorem corners_theorem (ε : ℝ) (hε : 0 < ε) :
+    ∃ n₀ : ℕ, ∀ G, [AddCommGroup G] → [Fintype G] → n₀ ≤ card G → ∀ A : Finset (G × G),
```

Comment:
There is an issue with the theorem as currently formulated that the `n₀` could depend on the universe (although obvisouly it doesn't). This is a general issue of theorems with an existential. There are two options here: either just formulate the theorem with `⌊(triangleRemovalBound (ε / 9) * 27)⁻¹⌋₊ + 1 ≤ card G`, or have a definition `def cornersTheoremBound (\epsilon) := ⌊(triangleRemovalBound (ε / 9) * 27)⁻¹⌋₊` and state the theorem with `cornersTheoremBound (\epsilon) ≤ card G`. I think I prefer the latter option.